### PR TITLE
Adds read-only support for containers

### LINKED ISSUE #16240: feature request: read-only containers
Docker 1.5 [introduced read-only containers](https://blog.docker.com/2015/02/docker-1-5-ipv6-support-read-only-containers-stats-named-dockerfiles-and-more/). I’ll quote from their release notes:

> By using this capability in combination with volumes, you can make sure containers only persist data where it can be managed in a known location.

I’d like to have kubernetes start all of my containers using `--read-only`, just so that I know there is no data which is stored in temporary storage and then lost when the container is restarted/rescheduled/…

Based on reading https://cloud.google.com/container-engine/docs/spec-schema, I don’t think kubernetes supports a read-only flag for the entire container, only for storage volumes. Hence, I’d like to suggest introducing a read-only flag for the entire container.

What do you think?

### PR DESCRIPTION
**Pull request by nickschuch**
Please refer to https://github.com/kubernetes/kubernetes/pull/17188 as the canonical source for "Read only container" support.

Fixes #16240

### CODE REVIEW

**Comment by smarterclayton** [comment]
Needs an e2e test

**Comment by smarterclayton** [comment]
I think this is probably too late to consider for 1.2 even though it is a very small change.  @kubernetes/kube-api would have to sign off on Monday.

I might recommend the name `ReadOnlyRootFilesystem` - there's no need to abbreviate in this case.

**Comment by smarterclayton** [comment]
@kubernetes/sig-node

**Comment by bgrant0607** [comment]
cc @dchen1107

**Comment by smarterclayton** [comment]
We have use cases for this in our high security production environments as one more control over abusive users, so it's a desirable feature.

**Comment by bgrant0607** [comment]
cc @vishh @thockin

**Comment by bgrant0607** [comment]
@smarterclayton Can you think of a reason to put this field into SecurityContext?

I'm fine with the feature and with the field name.

**Comment by smarterclayton** [comment]
I think it's related to security - it does seem like the primary reason you would do it as an admin would be to enforce it (via pod security policy) due to its implications.  As a developer, I think I would only do it because I wanted to test in that context before it going to that environment.  So I could buy the argument that it would be part of the container security context since other settings (like FSGroup, default permissions, volume security) would also be there.

**Comment by nickschuch** [comment]
```
Can you think of a reason to put this field into SecurityContext?
```

Is this something I should pursue?

```
Needs an e2e test
```

Worries, can you point me in the right direction for how I might write one?

```
ReadOnlyRootFilesystem
```

Happy to rename.

**Comment by smarterclayton** [comment]
I think so, yes - moving it to Security context and renaming it.

For the e2e test, typically you would add something to `test/e2e/` in one of the existing files (probably pods.go) and verifying that when you start a pod with this flag, if you try to exec into the container and write to a file, you'll get an error.  See `It("should contain environment variables for services [Conformance]", func() {` - you might start a container like that but with the command `echo 'test' > /etc/somefile` (probably pick a file that would be normally editable in that container) and check the returned log output to verify that the command failed with an appropriate error message, or by checking filesystem permissions.

**Comment by nickschuch** [comment]
Awesome! Ill get right on it.

I have completed the rename + the SecurityContext locally.

**Comment by nickschuch** [comment]
Ok, I have pushed:

* Renaming to ReadOnlyRootFilesystem
* Under the SecurityContext
* Added an e2e test

**Comment by nickschuch** [comment]
hmm, Im not sure how to fix the travis fails, I have rerun `hack/update-codecgen.sh` and see no changes. Also merged the latest master and reran. Any ideas?

**Comment by smarterclayton** [comment]
ok to test

**Comment by smarterclayton** [comment]
You need to update the swagger docs as well - be sure to run `hack/update-all.sh`

File: pkg/kubelet/container/runtime.go [hunk 1]
```diff
@@ -354,6 +354,8 @@ type RunContainerOptions struct {
 	DNSSearch []string
 	// The parent cgroup to pass to Docker
 	CgroupParent string
+	// The type of container rootfs
+	ReadOnly bool
```
Comment (vishh) [inline pkg/kubelet/container/runtime.go]: cc @yifan-gu

**Comment by vishh** [comment]
LGTM

**Comment by vishh** [comment]
@nickschuch: We are trying to move testing of node specific features to a separate node test suite. The migration is not complete yet. It will help if you can add a test for this feature [here](test/e2e_node/kubelet_test.go).

**Comment by nickschuch** [comment]
I have pushed the swagger documentation build. Will checkout that test. Will look at adding the test to the `/test/e2e_node/kubelet_test.go` file.

**Comment by nickschuch** [comment]
@vishh I have moved the test over to the newer test suite.

Im not sure how to resolve these fails:

```
-	for _, yyv74 := range v {
+	for _, yyv70 := range v {
```

Can I get some ideas on how to resolve? (first time Kubernetes contributor)

**Comment by smarterclayton** [comment]
Looks like node failure, will open an issue if it happens a second time

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue: #IGNORE

**Comment by vishh** [comment]
@nickschuch: That looks like generated code. Run `hack/update-all.sh`. That might just work.

**Comment by nickschuch** [comment]
@vishh ah! looks like my codecgen issue was related to running my kubernetes checkout in a different directory, moving to my GOPATH resolved.

@smarterclayton looks like it happened again, noooo! :)

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue: #IGNORE

**Comment by nickschuch** [comment]
Happy to keep rerolling this when required. Need anything from me or are we blocked by the bot?

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue: #IGNORE

It may end up being a problem with your pr, but I don't see anything obvious yetz

File: pkg/kubelet/dockertools/manager.go [hunk 2]
```diff
@@ -794,6 +795,11 @@ func usesHostNetwork(pod *api.Pod) bool {
 	return pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.HostNetwork
 }
 
+// determine if the container root should be a read only filesystem.
+func readOnlyRootFilesystem(container *api.Container) bool {
+	return container.SecurityContext != nil && *container.SecurityContext.ReadOnlyRootFilesystem
```
Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: Check for a nil read only rootfs value

Comment (smarterclayton) [inline pkg/kubelet/dockertools/manager.go]: That's probably causing the kubelet to panic, hence the e2e failures

**Comment by nickschuch** [comment]
Yeah, I was originally getting a panic here when referencing *container.SecurityContext.ReadOnlyRootFilesystem directly https://github.com/kubernetes/kubernetes/pull/20782/files#diff-0db82891d463ba14dd59da9c77f4776eR587

But then wrapping it in a check like this https://github.com/kubernetes/kubernetes/pull/20782/files#diff-0db82891d463ba14dd59da9c77f4776eR794 seemed to fix it.

Im able to run a pod with rootfs and no rootfs with those binaries.

Can't see anything of a panic in the Kubelet.log files in https://console.developers.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/20782/kubernetes-pull-build-test-e2e-gce/27940/

Gonna start at /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:190 and work through.

**Comment by nickschuch** [comment]
Got some GCE credit (trial) and running the `hack/e2e-test.sh`

**Comment by smarterclayton** [comment]
I don't see your fix, but if you mean you changed to check both SecurityContext != nil and ReadOnlyFilesystem != nil I don't see it in what's on GitHub.

**Comment by nickschuch** [comment]
Oh, must have just been on local. So I originally had this which resulted in panics on the Kubelet.

```
hc := &docker.HostConfig{
  PortBindings:   portBindings,
  Binds:          binds,
  NetworkMode:    netMode,
  IpcMode:        ipcMode,
  UTSMode:        utsMode,
  PidMode:        pidMode,
  ReadonlyRootfs: *container.SecurityContext.ReadOnlyRootFilesystem,
  // Memory and CPU are set here for newer versions of Docker (1.6+).
  Memory:     memoryLimit,
  MemorySwap: -1,
  CPUShares:  cpuShares,
}
```

I then migrated it over to this which fixed the issues.

```
hc := &docker.HostConfig{
  PortBindings:   portBindings,
  Binds:          binds,
  NetworkMode:    netMode,
  IpcMode:        ipcMode,
  UTSMode:        utsMode,
  PidMode:        pidMode,
  ReadonlyRootfs: readOnlyRootFilesystem(container),
  // Memory and CPU are set here for newer versions of Docker (1.6+).
  Memory:     memoryLimit,
  MemorySwap: -1,
  CPUShares:  cpuShares,
}

// determine if the container root should be a read only filesystem.
func readOnlyRootFilesystem(container *api.Container) bool {
	return container.SecurityContext != nil && *container.SecurityContext.ReadOnlyRootFilesystem
}
```

Rerunning the e2e from my GCE (ran into an issue) so I can try to replicate the current fails.

**Comment by smarterclayton** [comment]
Code looks ok, assuming the tests go green can you squash down to a single commit as per https://github.com/kubernetes/kubernetes/blob/d66ae29587e746c40390d61a1253a1bfa7aebd8a/docs/devel/development.md#when-to-retain-commits-and-when-to-squash

**Comment by nickschuch** [comment]
thanks @smarterclayton! thanks for bearing with me working through the problem. looks like it came down to the kube-proxy panicking which threw me out.

**Comment by smarterclayton** [comment]
```
Failure [10.185 seconds]
Generated release_1_2 clientset
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/generated_clientset.go:154
  should create pods, delete pods, watch pods [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/generated_clientset.go:153

  Expected
      <int>: 0
  to equal
      <int>: 1

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/generated_clientset.go:102
```

don't see evidence of a flake, but don't see an obvious failure.

@k8s-bot test this github issue: #IGNORE

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue: #IGNORE

**Comment by nickschuch** [comment]
We are further than we were before! Now we are getting proper build results.

**Comment by nickschuch** [comment]
Ah, looks like we have run into this #21045. Will keep and eye on it.

**Comment by smarterclayton** [comment]
@k8s-bot test this github issue: #21045

**Comment by nickschuch** [comment]
Looks like its closed off now, going to squash my commits.

**Comment by nickschuch** [comment]
Could have failed because of multiple tests running at the same time?

**Comment by nickschuch** [comment]
@k8s-bot test this github issue: #IGNORE

**Comment by erictune** [comment]
Now that PodSecurityPolicy is merged, should there be a corresponding field added to it?  @pweil- what do you think?

**Comment by pweil-** [comment]
seems like something an admin may want to control with policy if they're trying to prevent writing to local disks

**Comment by nickschuch** [comment]
```
Now that PodSecurityPolicy is merged, should there be a corresponding field added to it? @pweil- what do you think?
```

@erictune @pweil- can I get some details on what I need to do to for this?

**Comment by nickschuch** [comment]
Currently the ReadOnlyRootFilesystem is only available in "SecurityContext" and not "PodSecurityContext" so that would mean it doesn't adhere to PodSecurityPolicy? Or should this be in both?

**Comment by smarterclayton** [comment]
Having selective per container read only is useful in and of itself, it's potentially useful on pod security context too but we can apply the same rules there.  I think since this is a new change we don't need to block on enforcing read only for this PR, and can do a follow up to add that.

**Comment by pwittrock** [comment]
Looks like this broke the node e2e test suite:

>
09:00:58 I0216 09:00:58.908726   31269 run_e2e.go:103] Failure Finished Host e2e-node-coreos-beta.c.kubernetes-jenkins.internal Test Suite Failed to compile e2e_node:
09:00:58 
09:00:58 # k8s.io/kubernetes/test/e2e_node
09:00:58 test/e2e_node/kubelet_test.go:107: undefined: isReadOnly

**Comment by smarterclayton** [comment]
What path is that?

On Tue, Feb 16, 2016 at 12:57 PM, Phillip Wittrock <notifications@github.com

**Comment by smarterclayton** [comment]
Ah, I bet this was an in flight conflict.

**Comment by erictune** [comment]
Should add corresponding field to PodSecurityPolicy at some point.

**Comment by smarterclayton** [comment]
@pweil- has either already done it or is close to it - we merged the
control onto SCC in Origin 1.2

**Comment by carsonoid** [comment]
So am I correctly interpreting the state of readonly containers at this moment as technically supported by the kubelet but not yet allowed by the api?

**Comment by smarterclayton** [comment]
Readonly has been supported since 1.3 in the API.

**Comment by carsonoid** [comment]
@smarterclayton Thanks! I see the problem. I was trying to add it at the pod security context level and not per-container.

**Comment by carsonoid** [comment]
Oh yes I see it now. readOnlyRootFilesystem under securityContext. Thanks!

**Pull request merged**